### PR TITLE
Refactoring of ApplePay component 

### PR DESCRIPTION
**Pull request by neelSharma12**
### Chore
<chore>

-   Refactored ApplePay component in demo app.

</chore>

### CODE REVIEW

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        let request = PaymentDetailsRequest(
+            details: data.details,
+            paymentData: data.paymentData,
+            merchantAccount: ConfigurationConstants.current.merchantAccount
+        )
+        apiClient.perform(request) { [weak self] result in
+            self?.paymentResponseHandler(result: result)
+        }
+    }
+}
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:186]: No need for `ActionComponentDelegate` for Apple Pay

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    private func present(_ component: PresentableComponent) {
+        guard component.requiresModalPresentation else {
+            presenter?.present(viewController: component.viewController, completion: nil)
+            return
+        }
+
+        let navigation = UINavigationController(rootViewController: component.viewController)
+        component.viewController.navigationItem.rightBarButtonItem = .init(barButtonSystemItem: .cancel,
+                                                                           target: self,
+                                                                           action: #selector(cancelPressed))
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:139]: No need for this: we know exactly what is going to be presented :)

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+}
+
+extension ApplePayComponentAdvancedFlowExample: PaymentComponentDelegate {
+
+    internal func didSubmit(_ data: PaymentComponentData, from component: PaymentComponent) {
+        let request = PaymentsRequest(data: data)
+        apiClient.perform(request) { [weak self] result in
+            self?.paymentResponseHandler(result: result)
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:130]: There is no Action, so we can resolve `result` directly:
```suggestion
        switch result {
        case let .success(response):
            finish(with: response)
        case let .failure(error):
            finish(with: error)
        }
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        guard component.requiresModalPresentation else {
+            presenter?.present(viewController: component.viewController, completion: nil)
+            return
+        }
+
+        let navigation = UINavigationController(rootViewController: component.viewController)
+        component.viewController.navigationItem.rightBarButtonItem = .init(barButtonSystemItem: .cancel,
+                                                                           target: self,
+                                                                           action: #selector(cancelPressed))
+        presenter?.present(viewController: navigation, completion: nil)
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: No need for this: we know exactly what is going to be presented :)

```suggestion
        presenter?.present(viewController: component.viewController, completion: nil)
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 1 above)
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: No need for ActionComponentDelegate for Apple Pay:

```suggestion
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    // MARK: - Action Handling
+
+    internal lazy var adyenActionComponent: AdyenActionComponent = {
+        let handler = AdyenActionComponent(context: context)
+        handler.configuration.threeDS.delegateAuthentication = ConfigurationConstants.delegatedAuthenticationConfigurations
+        handler.configuration.threeDS.requestorAppURL = URL(string: ConfigurationConstants.returnUrl)
+        handler.delegate = self
+        handler.presentationDelegate = self
+        return handler
+    }()
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: No need for action Handling on Apple Pay :)

```suggestion
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    internal func applePayComponent(from paymentMethods: PaymentMethods?) -> ApplePayComponent? {
+        guard
+            let paymentMethod = paymentMethods?.paymentMethod(ofType: ApplePayPaymentMethod.self),
+            let applePayPayment = try? ApplePayPayment(payment: ConfigurationConstants.current.payment,
+                                                       brand: ConfigurationConstants.appName)
+        else { return nil }
+        var config = ApplePayComponent.Configuration(payment: applePayPayment,
+                                                     merchantIdentifier: ConfigurationConstants.applePayMerchantIdentifier)
+        config.allowOnboarding = true
+        config.supportsCouponCode = true
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: This property changes `amount`, hence doesn't work with Sessions 🤔

Maybe we can add a comment about it ?

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        config.shippingType = .delivery
+        config.requiredShippingContactFields = [.postalAddress]
+        config.requiredBillingContactFields = [.postalAddress]
+        config.shippingMethods = ConfigurationConstants.shippingMethods
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: This property changes amount, hence doesn't work with Sessions 🤔

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        guard component.requiresModalPresentation else {
+            presenter?.present(viewController: component.viewController, completion: nil)
+            return
+        }
+
+        let navigation = UINavigationController(rootViewController: component.viewController)
+        component.viewController.navigationItem.rightBarButtonItem = .init(barButtonSystemItem: .cancel,
+                                                                           target: self,
+                                                                           action: #selector(cancelPressed))
+        presenter?.present(viewController: navigation, completion: nil)
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]:
```suggestion
        presenter?.present(viewController: component.viewController, completion: nil)
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 3 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:130]: Resolved. 47a1fb69f153def70200c018673e639d6796d548

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 4 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 47a1fb69f153def70200c018673e639d6796d548

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 1 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 47a1fb69f153def70200c018673e639d6796d548

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift (hunk 6 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: 01dfeb7e415ba907570c43308df9830ec1cfa89e

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift (hunk 7 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: 01dfeb7e415ba907570c43308df9830ec1cfa89e

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift (hunk 8 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: 47a1fb69f153def70200c018673e639d6796d548

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 5 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 47a1fb69f153def70200c018673e639d6796d548

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    func didUpdate(contact: PKContact,
+                   for payment: ApplePayPayment,
+                   completion: @escaping (PKPaymentRequestShippingContactUpdate) -> Void) {
+        var items = payment.summaryItems
+        print(items.reduce("> ") { $0 + "| \($1.label): \($1.amount.floatValue.rounded()) " })
+        if let last = items.last {
+            items = items.dropLast()
+            let cityLabel = contact.postalAddress?.city ?? "Somewhere"
+            items.append(.init(label: "Shipping \(cityLabel)",
+                               amount: NSDecimalNumber(value: 5.0)))
```
Comment (goergisn) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:151]: Is this supposed to be a hardcoded value?

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+    @available(iOS 15.0, *)
+    func didUpdate(couponCode: String,
+                   for payment: ApplePayPayment,
+                   completion: @escaping (PKPaymentRequestCouponCodeUpdate) -> Void) {
+        var items = payment.summaryItems
+        print(items.reduce("> ") { $0 + "| \($1.label): \($1.amount.floatValue.rounded()) " })
+        if let last = items.last {
+            items = items.dropLast()
+            items.append(.init(label: "Coupon", amount: NSDecimalNumber(value: -5.0)))
```
Comment (goergisn) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:178]: Is this supposed to be a hardcoded value?

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        completion(.init(paymentSummaryItems: items))
+    }
+
+    @available(iOS 15.0, *)
+    func didUpdate(couponCode: String,
+                   for payment: ApplePayPayment,
+                   completion: @escaping (PKPaymentRequestCouponCodeUpdate) -> Void) {
+        var items = payment.summaryItems
+        print(items.reduce("> ") { $0 + "| \($1.label): \($1.amount.floatValue.rounded()) " })
```
Comment (goergisn) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: Is this debug print statement needed still?

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+            items.append(.init(label: last.label, amount: NSDecimalNumber(value: last.amount.floatValue + 5.0)))
+        }
+        completion(.init(paymentSummaryItems: items))
+    }
+
+    func didUpdate(shippingMethod: PKShippingMethod,
+                   for payment: ApplePayPayment,
+                   completion: @escaping (PKPaymentRequestShippingMethodUpdate) -> Void) {
+        var items = payment.summaryItems
+        print(items.reduce("> ") { $0 + "| \($1.label): \($1.amount.floatValue.rounded()) " })
```
Comment (goergisn) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: Is this debug print statement needed still?

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    func didFail(with error: Error, from component: Component, session: AdyenSession) {
+        dismissAndShowAlert(false, error.localizedDescription)
+    }
+
+    func didOpenExternalApplication(component: ActionComponent, session: AdyenSession) {}
+
+}
+
+extension ApplePayComponentExample: PresentationDelegate {
+    internal func present(component: PresentableComponent) {}
```
Comment (goergisn) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift:128]: Maybe good to add a comment here as well that the implementation of this delegate method is not needed

```suggestion
    internal func present(component: PresentableComponent) {
        // The implementation of this delegate method is not needed when using AdyenSession
    }
```

Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift:128]: 90ede7f5ceb40b17be5ba7e0cc83f1b617a6e213

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 12 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 930bacee2331858be5473647d4406808c9d72717

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 11 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 930bacee2331858be5473647d4406808c9d72717

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 10 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:178]: I have added a comment for now. 930bacee2331858be5473647d4406808c9d72717

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 9 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift:151]: I have added a comment for now https://github.com/Adyen/adyen-ios/commit/930bacee2331858be5473647d4406808c9d72717

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+            self.presentAlert(with: error)
+        }
+    }
+
+    internal func applePayComponent(from paymentMethods: PaymentMethods?) throws -> ApplePayComponent? {
+        guard
+            let paymentMethod = paymentMethods?.paymentMethod(ofType: ApplePayPaymentMethod.self),
+            let applePayPayment = try? ApplePayPayment(payment: ConfigurationConstants.current.payment,
+                                                       brand: ConfigurationConstants.appName)
+        else { throw IntegrationError.paymentMethodNotAvailable(paymentMethod: CardPaymentMethod.self)
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]:
```suggestion
        else { throw IntegrationError.paymentMethodNotAvailable(paymentMethod: ApplePayPaymentMethod.self)
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+        var config = ApplePayComponent.Configuration(payment: applePayPayment,
+                                                     merchantIdentifier: ConfigurationConstants.applePayMerchantIdentifier)
+        config.allowOnboarding = true
+        config.supportsCouponCode = true
+        config.shippingType = .delivery
+        config.requiredShippingContactFields = [.postalAddress]
+        config.requiredBillingContactFields = [.postalAddress]
+        config.shippingMethods = ConfigurationConstants.shippingMethods
+
+        let component = try? ApplePayComponent(paymentMethod: paymentMethod,
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]:
```suggestion
        let component = try ApplePayComponent(paymentMethod: paymentMethod,
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+            let component = try applePayComponent(from: paymentMethods)
+            guard let componentViewController = component?.viewController else { return }
+            presenter?.present(viewController: componentViewController, completion: nil)
+            applePayComponent = component
+        } catch {
+            self.presentAlert(with: error)
+        }
+    }
+
+    internal func applePayComponent(from paymentMethods: PaymentMethods?) throws -> ApplePayComponent? {
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]:
```suggestion
    internal func applePayComponent(from paymentMethods: PaymentMethods?) throws -> ApplePayComponent {
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+        guard
+            let paymentMethod = paymentMethods?.paymentMethod(ofType: ApplePayPaymentMethod.self),
+            let applePayPayment = try? ApplePayPayment(payment: ConfigurationConstants.current.payment,
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]:
```suggestion
            let applePayPayment = try ApplePayPayment(payment: ConfigurationConstants.current.payment,
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+            let component = try applePayComponent(from: session)
+            guard let componentViewController = component?.viewController else { return }
+            presenter?.present(viewController: componentViewController, completion: nil)
+            applePayComponent = component
+        } catch {
+            self.presentAlert(with: error)
+        }
+    }
+
+    internal func applePayComponent(from session: AdyenSession) throws -> ApplePayComponent? {
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]:
```suggestion
    internal func applePayComponent(from session: AdyenSession) throws -> ApplePayComponent {
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+        let paymentMethods = session.sessionContext.paymentMethods
+        guard let paymentMethod = paymentMethods.paymentMethod(ofType: ApplePayPaymentMethod.self),
+              let applePayPayment = try? ApplePayPayment(payment: ConfigurationConstants.current.payment,
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]:
```suggestion
              let applePayPayment = try ApplePayPayment(payment: ConfigurationConstants.current.payment,
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+            throw IntegrationError.paymentMethodNotAvailable(paymentMethod: CardPaymentMethod.self)
+        }
+        var config = ApplePayComponent.Configuration(payment: applePayPayment,
+                                                     merchantIdentifier: ConfigurationConstants.applePayMerchantIdentifier)
+        config.allowOnboarding = true
+        config.shippingType = .delivery
+        config.requiredShippingContactFields = [.postalAddress]
+        config.requiredBillingContactFields = [.postalAddress]
+
+        let component = try? ApplePayComponent(paymentMethod: paymentMethod,
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]:
```suggestion
        let component = try ApplePayComponent(paymentMethod: paymentMethod,
```

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+
+}
+
+extension ApplePayComponentAdvancedFlowExample: PresentationDelegate {
+
+    internal func present(component: PresentableComponent) {
+        let componentViewController = component.viewController
+        presenter?.present(viewController: componentViewController, completion: nil)
+    }
+}
```
Comment (descorp) [inline Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]:
```suggestion
```

This is not needed for advanced flow? There is no sessions and there is no ActionComponent?

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 14 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: Ah!. I missed this while copying the error :(. 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 15 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 17 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift (hunk 18 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift (hunk 19 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift (hunk 20 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentExample.swift]: 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 21 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: yes, you are right :) 434daaef01d20f3f62437a7110b7baf93ccfb13f

File: Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift (hunk 16 above)
Comment (neelSharma12) [reply Demo/Common/IntegrationExamples/Components/ApplePayComponentAdvancedFlowExample.swift]: 434daaef01d20f3f62437a7110b7baf93ccfb13f

**Review by descorp — APPROVED** [review]

**Review by goergisn — APPROVED** [review]

**Review by descorp — APPROVED** [review]

**Pull request merged**